### PR TITLE
tabulated printing added

### PR DESCRIPTION
**Pull request by someparsa**
with a new option **`-t`** in the command, users can have a tabulated format of the updateable packages. Very similar to what we have in the original `pip` package with `python -m pip list --outdated`.

### CODE REVIEW

File: pip_review/__main__.py
Comment (jgonggrijp) [inline pip_review/__main__.py]: "Tabulated" and "tabular format" (note the single 'b') are roundabout ways to say "table". Easy does it, so all else being equal, I would just go "table":

```suggestion
        '--table', '-t', action='store_true', default=False,
        help='Print the outdated packages in table format')
```

*However*, we cannot use the `-t` short option (at least not without it being a breaking change), because `pip install` has a `-t` for `--target` which some users might be currently forwarding through pip-review. The other pull request used `-p`, which does not have this problem. That `p` stood for `--preview`, but you can also stick with the long option `--table`, which is less consistent with the short option, but does describe the feature better in my opinion.

File: pip_review/__main__.py
Comment (jgonggrijp) [inline pip_review/__main__.py]: This insertion looks like you copied the `else` clause below and then made modifications to sculpt the behavior into something more to your liking. Which is totally understandable! However, there are some problems with this:

- Copying existing code and then modifying always introduced the problem that you are repeating a pattern. If you change the pattern in one place at some later time, you have to choose between updating all other copies of the pattern as well, or accepting that your program has become less consistent. The latter option is often a mistake, if you are changing the pattern because of a bug. Keep it DRY: Don't Repeat Yourself! (Or somebody else!)
- However, in this case I think the pattern you copied does not suit your goal very well, anyway. If `args.interactive` is true, the interactive prompts will interlace with the table rows, ruining your beautiful table. If somebody passes `--table --interactive` on the command line, I think they will rather expect `pip-review` to *first* print out the entire table and then *also* ask every package one by one, as is currently happening in the `else` clause.
- In addition, the pattern you copied did not need to align subsequent lines, so it could do everything in one loop. By copying this single loop, you forced yourself to hard-code the column widths, which is not great. Most of the time you are reserving too much space, while sometimes a long package name will still derail things. It would be much better to base the columns widths one the widest values in the actual data.

**Review by jgonggrijp — CHANGES_REQUESTED** [review]
Well done on creating a pull request!

Roughly the same feature was proposed before by @kunif in #91. Unfortunately, his work is unfinished, but there are some things he did right:

- He used `-p` as a short option instead of `-t`. `-p` has the advantage that it doesn't conflict with existing options of `pip list` and `pip install`.
- He dynamically sized the column widths, so the data always fit comfortably in the table.
- He ensured that the way the available updates are displayed (raw, wordy or tabular) was (mostly) independent of the way they are installed (automatically, interactively or not at all). This is not the case in the current version of your proposal: you cannot have a table and automatic installation at the same time.
- He documented the new feature in the README.

I suggest you read and learn from #91, including the comments, and also reuse the work that was already done over there, including the example code that I wrote in the comments. One particular way in which you might approach this, is as follows:

1. Add @kunif's fork as a remote to your local clone.<br>`git remote add -f kunif https://github.com/kunif/pip-review.git`
2. Create a new branch, based on @kunif's branch. This enables you to build on top of their hard work. Also, when your pull request is accepted, @kunif will automatically be credited as well.<br>`git checkout -b table-option kunif/add-preview-options`
3. Make the changes I suggested in #91. You are welcome to just blatantly copy the code I wrote. Commit these changes before continuing.
4. Make any other changes you want to make. Maybe you want to change `--preview` into `--table`. Maybe you can find a way so that the extra `-P` option is not needed for people who want a table and no installation (arguably, this is what should happen by default when only the `-p` option is passed, which is also how your current proposal behaves). Commit at least once along the way.
5. Push your `table-option` branch to your fork.
6. Update the current pull request to request that your `table-option` branch is merged instead of your `develop` branch. Failing that (if the option is not available, or it does not work for whatever reason), submit a new pull request instead.

Some more comments below. Good luck and please let me know if you have any questions!

**Comment by someparsa** [comment]
Thank you Julian, I will gradually read the text and progress. Will report when I reach any good point. Sorry, maybe my speed is a little slow. Many thanks again :)

**Comment by jgonggrijp** [comment]
No worries, take your time!

**Comment by kunif** [comment]
By the way, most of my work was left unneeded because the reviews wiped out the traces. However, should my pull request be completed with the review results reflected?

**Comment by jgonggrijp** [comment]
@kunif Yes, I think your pull request should be completed. This is exactly what will happen if @someparsa follows my instructions. Please let us know if you want to complete it yourself; in that case, @someparsa can focus his attention on other features.

> most of my work was left unneeded because the reviews wiped out the traces.

In my comments, I suggested changes that would replace most of your code by other code, but your work was still needed in order to arrive at those suggestions. A contribution does not count because its code is retained byte for byte, but because it is a step in the evolution of the product. Eventually, every byte of code will be replaced, but all contributions along the way still count.

**Comment by someparsa** [comment]
Hi Julian @jgonggrijp and Kunio @kunif, I am working gradually on the version you guys summed up in your commit together. It is nice and I am gradually finding my way. I think that there is a capacity to improve the code.

We get a list of outdated packages with this.

`
outdated = get_outdated_packages(list_args)
`

However, any calling in our code, reading existing list of packages seems to be time consuming. I think it will be an advantage if we code something or alter our methods to decrease the time needed reading the packages on the system. This is also occupying my mind now, along with our previous discussion on tabulating the info.

**Comment by someparsa** [comment]
I think that the whole functionality of the package can be categorized in the following patterns. Therefore I think we can define these `args` for the package (probably rename the existing). My suggestions in the brackets [].

#### a- to preview the updateable packages
  1. in raw format [**`--preview-raw -p`**]
  2. in table format [**`--preview-table -t`**]

#### b- to update the updateable packages
  1. automatically [**`--update-auto -a`**]
  2. selectively / interactively [**`--update-selection -s`**]
  3. abort [**`--abort -a`**]

And may be there is no need to mix the preview command and update command. The user can preview the updateable in any format they like. And they can do the update in any procedure it is convenient. The only case of concern would be the **`--update-selection`** case. We may decide in which format we should show the packages in the interactive selection case. In this case the most convenient would be the raw format. Otherwise we can give the chance to the user to select how they want to see updateable packages in each step, in table or raw format.

  1. selectively - raw [**`--update-selection-raw -s`**] (small **`s`**)
  2. selectively - table [**`--update-selection-table -S`**] (capital **`S`**)

Coding such classification will make development much more convenient.

**Comment by jgonggrijp** [comment]
@someparsa 

> Hi Julian and Kunio, I am working gradually on the version you guys summed up in your commit together. It is nice and I am gradually finding my way. I think that there is a capacity to improve the code.

Great to hear that you feel inspired!

> 
> We get a list of outdated packages with this.
> 
> `outdated = get_outdated_packages(list_args)`
> 
> However, any calling in our code, reading existing list of packages seems to be time consuming. I think it will be an advantage if we code something or alter our methods to decrease the time needed reading the packages on the system. This is also occupying my mind now, along with our previous discussion on tabulating the info.

*Danger, Will Robinson!*

Which of the following scenarios applies to your suspicion?

1. You used `pip-review` to manage your dependencies and noticed that it was slow. You used a profiler to determine what was taking so much time. You found that most time was spent in one or more calls to `get_outdated_packages`.
2. You read the code and noticed the calls to `get_oudated_packages` and the implementation of `get_outdated_packages`. It seemed to you that executing that code might be time-consuming.
    1. Following up, you spent some more time thinking about your observation. Then, you posted the above comment.
    2. Following up, you prepared several virtual environments with different sets of packages installed, some outdated. In each virtual environment, you ran `pip-review` several times, recording the total CPU time each time. You also made multiple measurements of the time required to run just `pip list --outdated`. You found that `pip-review` took significantly longer than `pip list --outdated` on average. Then, you posted the above comment.

If scenario 2.i. applies to you, this is what people refer to as "premature optimization": wanting to change code in order to improve performance, without solid evidence that there is a performance problem in the first place.

A famous computer scientist (Donald Knuth) said that premature optimization is the root of all evil. This is not true (there are other sources of evil as well), but you still should not optimize code without solid evidence that there is a problem. The optimized version of a piece of code generally takes up more space and has more details than the original code. This means there are also more opportunities for bugs, so you have a larger *maintenance surface* to deal with. Moreover, if the optimization is not evidence-based, you cannot be sure that the performance will improve. It might actually get worse!

That said, if scenario 2.ii. applies to you, or even better, scenario 1., then please bring forward the evidence!

> I think that the whole functionality of the package can be categorized in the following patterns. Therefore I think we can define these `args` for the package (probably rename the existing). My suggestions in the brackets [].

Better not to rename the flags, because that will break people's scripts in which they automatically call `pip-review`. Apart from that, though, I agree that the preview format and the update mode should just vary independently from each other.

> 
> #### a- to preview the updateable packages
> 1. in raw format [**`--preview-raw -p`**]
> 2. in table format [**`--preview-table -t`**]

There is a third mode already in existence, which is the default if you don't pass `--raw` (or `--preview`/`--preview-table` once that is added). It looks a bit like the raw format but embedded in an English sentence that also states the old version. I previously referred to this format as the "wordy" format.

> 
> #### b- to update the updateable packages
> 1. automatically [**`--update-auto -a`**]
> 2. selectively / interactively [**`--update-selection -s`**]
> 3. abort [**`--abort -a`**]

Abort is just what happens when you don't pass `--auto` or `--interactive`, so it does not need a separate flag.

> 
> And may be there is no need to mix the preview command and update command. The user can preview the updateable in any format they like. And they can do the update in any procedure it is convenient.

Yes, that's how it should be.

> The only case of concern would be the **`--update-selection`** case. We may decide in which format we should show the packages in the interactive selection case. In this case the most convenient would be the raw format. Otherwise we can give the chance to the user to select how they want to see updateable packages in each step, in table or raw format.
> 
> 1. selectively - raw [**`--update-selection-raw -s`**] (small **`s`**)
> 2. selectively - table [**`--update-selection-table -S`**] (capital **`S`**)
> 
> Coding such classification will make development much more convenient.

I think the interactive selection should always use the wordy format. If the user also passes the `--raw` or `--preview` flag, then they should just get the entire list of outdated packages in that format first, before the interactive selection begins.

By the way, I think @kunif decided he wants to finish #91 anyway. I suggest we let him finish that job, and then you can add your refinements on top.

**Comment by someparsa** [comment]
Hi Julian

My understanding is a mixture of two scenarios you mentioned above, although I did not make a measurement to understand the root of the time demand. My understanding is that, at the present we should polish the interface, let's say the user experience, then later we can look back to the back-end code and probably improve them. Specifically last night I was getting lost between the functions processing things. So just simplified my focus by digesting how things appear on the shell. This will increase my speed dealing with this package. So I will give priority to the way the package interacts in the command window.

About your point of view on not changing the existing tags, I agree. It will confuse the users so let's stick to what we have and possibly add new ones on demands. BTW I think that the separation of the procedures for the previews and processes can be a good improvement. All my thinking here is how easier the user interacts with the package. This can happen by clear logs, may be progress bars as we read in other comments, etc.

Yes please, let's finalize @kunif work and then we continue building over the newer release. This makes working with the packages much easier as I could not make a new fork from Kunio's source (duplicate source forks not allowed?) Maybe I should learn how to fork a forked source in a minor branch, when I have already forked the main source. I should learn to make separate branches to develop features in different roots.

And finally, I am so impressed by seeing the way you comment and spend time for the developers. Good amount of text, deep information and practical step by step advices. What else do we want from a team developing open-source? Thanks indeed, you are so encouraging and giving confidence to me.

Also, reading about Donald Knuth was also lovely. His work on TeX is great. It is a long time I am using LaTeX for my work, thesis, CV, etc. That is awesome.

Looking forward seeing the progress on other branch with Kunif and then I will step in to add some lines of code.

**Comment by jgonggrijp** [comment]
@someparsa 

> (...) My understanding is that, at the present we should polish the interface, let's say the user experience, then later we can look back to the back-end code and probably improve them.

Agreed.

> Specifically last night I was getting lost between the functions processing things. So just simplified my focus by digesting how things appear on the shell. This will increase my speed dealing with this package. So I will give priority to the way the package interacts in the command window.

Good.

> 
> (...) BTW I think that the separation of the procedures for the previews and processes can be a good improvement. All my thinking here is how easier the user interacts with the package. This can happen by clear logs, may be progress bars as we read in other comments, etc.
> 
> Yes please, let's finalize @kunif work and then we continue building over the newer release.

Right. I think after merging in @kunif's changes, I will let you adjust the flags first before publishing another release. That way, we avoid introducing a `--preview-only` tag which immediately disappears again in the next release.

> This makes working with the packages much easier as I could not make a new fork from Kunio's source (duplicate source forks not allowed?)

Ah right, duplicate forks are impossible. But you don't need them, because your local clone can be linked to my public repository as well as your own fork *and* @kunif's fork, all at the same time. (What is a "local clone", you ask? We will get to that next.)

> Maybe I should learn how to fork a forked source in a minor branch, when I have already forked the main source. I should learn to make separate branches to develop features in different roots.

Yes, I think you need to get more familiar with Git. In fact, I suggest you do that first, while you wait for me to merge in @kunif's changes. Two options (both available free of charge online):

1. [Git Immersion](https://gitimmersion.com) will get you up to speed quickly. It does not teach you all the details, but you'll know enough for your next contribution by the end of the tutorial.
2. [Pro Git](https://git-scm.com/book/en/v2) will teach you all the details. It is an entire book, so this will take longer, but I found it worth the effort.

You can also do Git Immersion first and save Pro Git for later. That's what I did.

Also, if you run into any hurdles while working through one of these tutorials, please feel welcome to ask for help here.

> 
> And finally, I am so impressed by seeing the way you comment and spend time for the developers. Good amount of text, deep information and practical step by step advices. What else do we want from a team developing open-source? Thanks indeed, you are so encouraging and giving confidence to me.

Glad to hear that!

**Comment by jgonggrijp** [comment]
@someparsa I just merged #91, and I suspect you made some progress learning Git. This might be a good time to discuss some next steps.

I'm going to close this pull request, because the changes in here are redundant with #91. **However**, I propose we keep using this ticket to discuss your work until you open the next pull request.

Here is a draft of plan. It is up to you whether you want to follow it exactly, or make adjustments:

1. Finish any Git training material you're currently working through.
2. If you have not already done so, make a local clone of your own fork of pip-review so you can pull from and push to it.

```sh
cd /a/place/on/disk/where/you/store/projects
# SSH variant (based on a public key that you need to generate and upload to GitHub first):
git clone git@github.com:someparsa/pip-review.git
# HTTPS variant (based on a password):
git clone https://github.com/someparsa/pip-review.git
cd pip-review
# your fork is now the "origin" remote
```

3. Add my public repository as the "upstream" remote, using the HTTPS URL, so you can pull from this repository as well.

```sh
git remote add -f upstream https://github.com/jgonggrijp/pip-review.git
```

4. Optionally add more remotes, for example Kunio's fork.

```sh
git remote add -f kunif https://github.com/kunif/pip-review.git
```

5. Optionally create a backup of your current `develop` branch.

```sh
git branch tabulated-printing-bak develop
```

6. Reset your local `develop` and your public fork's `develop` so they are identical to the `develop` on my repository. This means the changes you made will disappear from those branches, but this is necessary in order to keep things tidy in the future.

```sh
git checkout develop
git reset --hard upstream/develop
git push --force origin develop
```

7. Fine-tune your configuration so that your local clone will *pull* `develop` and `master` from _my_ repository by default, but *push* them to _your_ own fork by default. Again, this will help to keep things tidy in the future.

```sh
git branch master origin/master # only needed if you have no local master yet
git config remote.pushDefault origin
git config branch.master.remote upstream
git config branch.develop.remote upstream
```

8. Ideally, install the [git-flow](https://github.com/nvie/gitflow#readme) helper tool, then configure it in your local clone.

```sh
git flow init # all default values are fine
```

9. Decide what you're going to work on first, then create a separate branch for it. I'm using the name `smooth-options` below, on the assumption that you're first going to streamline the options as discussed above.

```sh
# make sure you work on top of the latest code first
git checkout develop
git pull
# start a new branch, two options:
# either (1) if you installed git-flow
git flow feature start smooth-options
# OR (2) otherwise
git checkout -b feature/smooth-options
```

10. Make your changes, committing along the way. If you're streamlining the options, I think your changes will consist of (a) removing the `--preview-only` option and (b) changing the `if`/`else` structure in `main` so the remaining options behave as intended.
11. Push your feature branch to your fork and create a new pull request.

```sh
git push # after this command,
# GitHub will helpfully show a URL that you can visit in order to create a PR
```

Good luck! Please comment here if you have any questions or comments.

**Comment by someparsa** [comment]
Hi Julian, Wow! every time I wonder and get surprised with how you kindly train me with your comments. Many thanks and I wish one day I can do something good for you in turn. Anyway, I will progress with these and will inform you later. About ssh and gpg keys, yes I have set them up on my account. They are great. Currently I can do basic tasks like git add, git commit, git push, etc. I use these on my own programming in the past two months and using git have significantly improved my coding quality. Though I have some difficulty with more advanced features of git. The world becomes dark when conflicts appear between commits and I can hardly handle them. :)) I am working on these. I will update you with the new release of the package and my readings on git. stay well.

**Pull request closed without merging**
